File: frequency_counter/same_frequency.py
# Write a function called same_frequency. Given two positive integers, find out if the two numbers have the same frequency of digits. The solution must have a time complexity of O(N)


def same_frequency(num1, num2):
    num1_frequencies = convert_to_frequencies(num1)

    for digit in str(num2):
        if digit not in num1_frequencies or num1_frequencies[digit] == 0:
            return False
        else:
            num1_frequencies[digit] -= 1

    return True


def convert_to_frequencies(num):
    frequencies = {}

    for digit in str(num):
        frequencies[digit] = frequencies[digit] + 1 if digit in frequencies else 1
        
    return frequencies


print(same_frequency(182, 281))  # true
print(same_frequency(122333, 333221))  # true
print(same_frequency(122333, 111223))  # false
print(same_frequency(34, 14))  # false
print(same_frequency(12345, 53241))  # true
print(same_frequency(22, 222))  # false

File: frequency_counter/sameFrequency.ts
function sameFrequency(num1: number, num2: number): boolean {
	const num1Frequencies: any = {};

	for (const digit of String(num1)) {
		num1Frequencies[digit] = (num1Frequencies[digit] || 0) + 1;
	}

	for (const digit of String(num2)) {
		if (!(digit in num1Frequencies) || num1Frequencies[digit] === 0) {
			return false;
		} else {
			num1Frequencies[digit] -= 1;
		}
	}

	return true;
}

console.log(sameFrequency(182, 281)); // true
console.log(sameFrequency(122333, 333221)); // true
console.log(sameFrequency(122333, 111223)); // false
console.log(sameFrequency(34, 14)); // false
console.log(sameFrequency(12345, 53241)); // true
console.log(sameFrequency(22, 222)); // false
